src/coreclr/vm/profdetach.cpp

```diff
@@ -467,27 +482,35 @@ void ProfilingAPIDetach::UnloadProfiler()
     }
     CONTRACTL_END;
 
-    _ASSERTE(g_profControlBlock.curProfStatus.Get() == kProfStatusDetaching);
+    _ASSERTE(pDetachInfo->m_pProfilerInfo->curProfStatus.Get() == kProfStatusDetaching);
 
     {
         CRITSEC_Holder csh(ProfilingAPIUtility::GetStatusCrst());
 
         // Notify profiler it's about to be unloaded
-        _ASSERTE(s_profilerDetachInfo.m_pEEToProf != NULL);
-        s_profilerDetachInfo.m_pEEToProf->ProfilerDetachSucceeded();
-
-        // Reset detach state.
-        s_profilerDetachInfo.Init();
+        _ASSERTE(pDetachInfo->m_pProfilerInfo != NULL);
+        
+        {
+            EvacuationCounterHolder evacuationCounter(pDetachInfo->m_pProfilerInfo);
```

Comment:
It is not necessary for correctness, it is only there because we assert in EEToProfInterfaceImpl that if we're in a callback then the evacuation counter must be >0. I'll add a comment